### PR TITLE
Lora upload & Upload enhancements

### PR DESCRIPTION
**Pull request by valaises**
### Features:
* Added reusable "Upload Files Modal" component
* Sources: replaced "upload files and upload link" with the component
* Finetune: added "Upload lora btn" that lets you upload your custom lora via link or file input

### CODE REVIEW

**Review by olegklimov — APPROVED** [review]
Good changes, but I didn't test yet.

**Comment by olegklimov** [comment]
@valaises change destination to `dev` pls

**Review by hazratisulton — CHANGES_REQUESTED** [review]
Tried upload by link even with set name manually:
image
image
Tried upload from local zip file - worked.

File: refact_data_pipeline/finetune/process_uploaded_files.py [hunk 1]
```diff
@@ -330,10 +330,11 @@ def mark_which_set(lst, subdir):
...
 
-def rm_and_unpack(upload_filename, unpack_filename, source_type, filename):
-    cmd = ["rm", "-rf", unpack_filename]
-    log(" ".join(cmd))
-    subprocess.check_call(cmd)
+def rm_and_unpack(upload_filename, unpack_filename, source_type, filename, rm_unpack_dir: bool = True):
+    if rm_unpack_dir:
+        cmd = ["rm", "-rf", unpack_filename]
+        log(" ".join(cmd))
+        subprocess.check_call(cmd)
```
Comment (olegklimov) [inline refact_data_pipeline/finetune/process_uploaded_files.py:337]: shutil

File: self_hosting_machinery/webgui/static/components/modals/modal-upload-files.js [hunk 2]
```diff
@@ -0,0 +1,364 @@
...
+                    btns_nav_upload_files.forEach(el => el.classList.remove('active', 'main-active'));
+                    el.classList.add('active', 'main-active');
+                    panes_upload_files.forEach(el => el.classList.remove('show', 'active'));
+                    file_modal.querySelector(`#${el.dataset.toggle}`).classList.add('show', 'active')
+                }
+            })
+        );
+    }
+
+    submit_event(submit_link_endpoint, submit_input_endpoint, text_on_progress_done) {
```
Comment (olegklimov) [inline self_hosting_machinery/webgui/static/components/modals/modal-upload-files.js:147]: add_submit_handler

File: self_hosting_machinery/webgui/static/components/modals/modal-upload-files.js [hunk 3]
```diff
@@ -0,0 +1,364 @@
...
+
+    hide_modal() {
+        bootstrap.Modal.getOrCreateInstance(document.getElementById('updlg-modal-upload-files')).hide();
+    }
+
+    show_modal() {
+        bootstrap.Modal.getOrCreateInstance(document.getElementById('updlg-modal-upload-files')).show();
+    }
+
+    process_now_update_until_finished(upload_method) {
```
Comment (olegklimov) [inline self_hosting_machinery/webgui/static/components/modals/modal-upload-files.js:216]: prepare_for_upload

File: self_hosting_machinery/webgui/static/components/modals/modal-upload-files.js [hunk 4]
```diff
@@ -0,0 +1,364 @@
...
+                this.reset_modal_fields();
+                return response.json();
+            }).then(
+                data => {
+                    this.reset_modal_fields();
+                    this.hide_modal();
+                    }).catch(
+                        error => {
+                            this.reset_modal_fields();
+                            file_modal.querySelector('#updlg-upload-files-status').innerHTML = error.message;
```
Comment (olegklimov) [inline self_hosting_machinery/webgui/static/components/modals/modal-upload-files.js:361]: innerText

File: self_hosting_machinery/webgui/static/components/modals/modal-upload-files.js [hunk 5]
```diff
@@ -0,0 +1,364 @@
...
+                if (!response.ok) {
+                    return response.json()
+                        .then(error => {
+                            throw new Error(error.message);
+                        });
+                }
+                this.reset_modal_fields();
+                return response.json();
+            }).then(
+                data => {
```
Comment (olegklimov) [inline self_hosting_machinery/webgui/static/components/modals/modal-upload-files.js:355]: indent

File: self_hosting_machinery/webgui/static/components/modals/modal-upload-files.js [hunk 6]
```diff
@@ -0,0 +1,364 @@
...
+                    this.reset_modal_fields();
+                    this.hide_modal();
+                    }).catch(
```
Comment (olegklimov) [inline self_hosting_machinery/webgui/static/components/modals/modal-upload-files.js:358]: check if it catches errors in both blocks of code

File: self_hosting_machinery/webgui/static/tab-upload.js [hunk 7]
```diff
@@ -737,10 +608,22 @@ export async function init(general_error) {
...
+        '/tab-files-upload-url', '/tab-files-upload',
+        "Wrapping up. Please wait...",
+        "https://yourserver.com/file.zip",
+        "You can upload .zip, .tar.gz, .tar.bz2 archives, or an individual file such as \"my_program.py\""
+    );
 }
 
 export function tab_switched_away() {
     dont_disable_file_types = false;
+    document.querySelector('#upload-files-modal').innerHTML = "";
```
Comment (olegklimov) [inline self_hosting_machinery/webgui/static/tab-upload.js:626]: symmetric

File: self_hosting_machinery/webgui/tab_loras.py [hunk 8]
```diff
@@ -0,0 +1,91 @@
...
+from self_hosting_machinery import env
+from self_hosting_machinery.webgui.selfhost_webutils import log
+from self_hosting_machinery.webgui.tab_upload import download_file_from_url, UploadViaURL
+
+
+def rm(f):
+    try:
+        subprocess.check_call(['rm', '-rf', f])
+    except BaseException as e:
+        log(f"Error while removing {f}: {e}")
```
Comment (olegklimov) [inline self_hosting_machinery/webgui/tab_loras.py:19]: shutil

File: self_hosting_machinery/webgui/tab_upload.py [hunk 9]
```diff
@@ -1,39 +1,67 @@
...
+                with open(file_path, 'wb') as file:
+                    async for chunk in response.content.iter_chunked(1024 * 1024):
+                        file.write(chunk)
+    except Exception as e:
+        try:
+            if file_path and os.path.exists(file_path):
+                os.remove(file_path)
+        except Exception:
+            log(f"Cannot remove file {file_path}")
+        raise e
```
Comment (olegklimov) [inline self_hosting_machinery/webgui/tab_upload.py:58]: raise

File: self_hosting_machinery/webgui/tab_upload.py [hunk 10]
```diff
@@ -169,15 +197,11 @@ async def _tab_files_upload(self, file: UploadFile):
...
-        last_path_element = os.path.split(post.url)[1]
-        file_path = os.path.join(env.DIR_UPLOADS, last_path_element)
         try:
-            with open(file_path, "wb") as f:
-                f.write(bin)
-        except OSError as e:
-            return JSONResponse({"message": f"Error: {e}"}, status_code=500)
+            await download_file_from_url(post.url, env.DIR_UPLOADS, post.filename)
+        except Exception as e:
+            return JSONResponse({"message": f"Cannot download: {e}"}, status_code=500)
```
Comment (olegklimov) [inline self_hosting_machinery/webgui/tab_upload.py:203]: "detail"

File: self_hosting_machinery/webgui/static/components/modals/modal-upload-files.js [hunk 11]
```diff
@@ -0,0 +1,364 @@
...
+    </div>
+  </div>
+</div>
+`
+
+let gl_open_on_click_el;
+let gl_open_on_click_el_default_html;
+let gl_modal;
+
+function makeModalBackdropStatic() {
```
Comment (olegklimov) [inline self_hosting_machinery/webgui/static/components/modals/modal-upload-files.js:63]: snake

File: self_hosting_machinery/webgui/tab_loras.py [hunk 12]
```diff
@@ -0,0 +1,91 @@
...
+
+
+async def unpack(file_path: Path) -> JSONResponse:
+    if not file_path.is_file():
+        return JSONResponse({"message": f"Error while unpacking: File {file_path.name} does not exist"}, status_code=404)
+
+    if get_source_type(str(file_path)) != 'archive':
+        return JSONResponse({"message": f"Error while unpacking: File {file_path.name} is not an archive; try setting filename manually"}, status_code=400)
```
Comment (olegklimov) [inline self_hosting_machinery/webgui/tab_loras.py:27]: "detail"

File: self_hosting_machinery/webgui/tab_loras.py [hunk 13]
```diff
@@ -0,0 +1,91 @@
...
+            rm(f)
+            return resp
+
+        return JSONResponse("OK", status_code=200)
+
+    async def _upload_lora_url(self, file: UploadViaURL):
+        try:
+            file_path = await download_file_from_url(file.url, env.DIR_LORAS, file.filename)
+        except Exception as e:
+            return JSONResponse({"message": f"Cannot download: {e}"}, status_code=500)
```
Comment (olegklimov) [inline self_hosting_machinery/webgui/tab_loras.py:85]: "detail"

**Comment by olegklimov** [comment]
`file xxxxxxx --mime-type --brief`

**Pull request merged**